### DISCUSSION TITLE
How to apply directives on input object fields

### DISCUSSION THREAD

**Discussion by de55** [ExpediaGroup/graphql-kotlin#1294] [General] [answered]
I have a field directive like this:

```
    override fun onField(environment: KotlinFieldDirectiveEnvironment): GraphQLFieldDefinition {
        val field = environment.element
        val originalDataFetcher: DataFetcher<*> = environment.getDataFetcher()
        val newFetcher = DataFetcher<Any> { dataEnv ->
            val fourDigitNumber: String? = dataEnv.getArgument(environment.element.arguments[0].name) as String?
            if (fourDigitNumber?.length != 4 || fourDigitNumber.toLongOrNull() == null) {
                throw Exception("Must only contain 4 numerical characters")
            }
            originalDataFetcher.get(dataEnv)
        }
        environment.setDataFetcher(newFetcher)
        return field
    }
```

This works great on arguments, but it does not work on an input object field. My first thought was to do the same thing overriding onInputObjectField, but the get/set datafetcher methods don't exist:

```
    override fun onInputObjectField(environment: KotlinSchemaDirectiveEnvironment<GraphQLInputObjectField>): GraphQLInputObjectField {
        val field = environment.element
        val originalDataFetcher: DataFetcher<*> = environment.getDataFetcher() //THIS METHOD DOESN'T EXIST
        val newFetcher = DataFetcher<Any> { dataEnv ->
            val fourDigitNumber: String? = dataEnv.getArgument(environment.element.arguments[0].name) as String?
            if (fourDigitNumber?.length != 4 || fourDigitNumber.toLongOrNull() == null) {
                throw Exception("Must only contain 4 numerical characters")
            }
            originalDataFetcher.get(dataEnv)
        }
        environment.setDataFetcher(newFetcher) //THIS METHOD DOESN'T EXIST
        return field
    }
```

Is there a way to do this on the fields of input objects? Or will this only work on arguments?

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
In order to access the data fetcher you need to have schema coordinates and they are only available on the field (i.e. if you define an object it could be return from many different places during schema execution). Each directive environment has access to the raw code registry so while you won't get the convenient function to access the data fetcher, you still should be able to locate it in the registry (added in https://github.com/ExpediaGroup/graphql-kotlin/pull/1159).

**Comment by de55** [comment]
Ahh that makes sense thanks. We are on an older version of the library, so codeRegistry wasn’t available, but we will look into upgrading now.

Thanks for the quick response?
________________________________
From: Dariusz Kuc ***@***.***>
Sent: Wednesday, October 27, 2021 5:33:14 PM
To: ExpediaGroup/graphql-kotlin ***@***.***>
Cc: Derek Towriss ***@***.***>; Author ***@***.***>
Subject: [External] Re: [ExpediaGroup/graphql-kotlin] How to apply directives on input object fields (Discussion #1294)


Hello 👋
In order to access the data fetcher you need to have schema coordinates and they are only available on the field (i.e. if you define an object it could be return from many different places during schema execution). Each directive environment has access to the raw code registry so while you won't get the convenient function to access the data fetcher, you still should be able to locate it in the registry (added in #1159<https://github.com/ExpediaGroup/graphql-kotlin/pull/1159>).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ExpediaGroup/graphql-kotlin/discussions/1294#discussioncomment-1548825>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMADE4NA36BRO74Z5E5K5SDUJBVZVANCNFSM5G3LB2KQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

This e-mail message (including attachments, if any) is intended only for the use of the individual or entity to which it is addressed and may contain information that is privileged, proprietary, confidential and exempt from disclosure. If you are not the intended recipient, any dissemination, distribution or copying of this communication is strictly prohibited. If you have received this communication in error, please notify the sender and delete this e-mail message immediately.
